Add access to dense_vector values

mayya-sharipova opened the pull request:
Allow direct access to a dense_vector' values in script
through the following functions:

- getVectorValue – returns a vector's value as an array of floats
- getVectorMagnitude – returns a vector's magnitude

Closes #51964

mayya-sharipova committed:
Add access to dense_vector values

Allow direct access to a dense_vector' values in script
through the following functions:

- getVectorValue – returns a vector's value as an array of floats
- getVectorMagnitude – returns a vector's magnitude

Closes #51964

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova committed:
Small corrections

mayya-sharipova commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/mapper/VectorEncoderDecoder.java:
```diff
@@ -31,7 +32,39 @@ public static int denseVectorLength(Version indexVersion, BytesRef vectorBR) {
+            ((vectorBR.bytes[offset+2] & 0xFF) <<  8) |
+            (vectorBR.bytes[offset+3] & 0xFF);
+        return Float.intBitsToFloat(intValue);
+    }
```
I removed using ByteBuffer while retrieving a vector's magnitude. First the benchmarks shows that the performance is almost the same, second we don't need to create extra ByteBuffer objects that would need be garbage collected later.

10 dims

```
Benchmark                                    Mode  Cnt  Score   Error  Units
VectorFunctionBenchmark.decodeMag            avgt   30  3.953 ± 0.101  ns/op
VectorFunctionBenchmark.decodeMagWithBuffer  avgt   30  3.504 ± 0.053  ns/op
```

100 dims
```
VectorFunctionBenchmark.decodeMag            avgt   30  3.728 ± 0.103  ns/op
VectorFunctionBenchmark.decodeMagWithBuffer  avgt   30  3.276 ± 0.067  ns/op
```

512 dims
```
VectorFunctionBenchmark.decodeMag            avgt   30  3.752 ± 0.099  ns/op
VectorFunctionBenchmark.decodeMagWithBuffer  avgt   30  3.377 ± 0.023  ns/op
```

mayya-sharipova replied:
@jtibshirani Thanks for the feedback. I agree to keep the PR focused, and I removed this change from this PR. But  may be we can follow it up later.    I wonder what is the implication of creating and garbage collecting of a `ByteBuffer` object, is it work extra `0.5ns`?

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into dense-vector-iterator

jtibshirani commented on docs/reference/vectors/vector-functions.asciidoc:
```diff
@@ -8,6 +8,16 @@ linearly scanned. Thus, expect the query time grow linearly
+2. `dotProduct` – calculates dot product
+3. `l1norm` – calculates L^1^ distance
+4. `l2norm` - calculates L^2^ distance
+5. `getVectorValue` – returns a vector's value as an array of floats
```
It looks like painless will expose these methods as `doc['field'].vectorValue` and `doc['field'].vectorMagnitude`. Should we emphasize that usage pattern instead of calling getters directly? Also I wonder if we could rename `vectorMagnitude` to `magnitude` for simplicity.

mayya-sharipova replied:
That's very correct,  using just `vectorValue` and `magnitude` access methods are much better. Addressed in 388b1f85029214d3b8cd53f5f2ed5aec2023f6f3

jtibshirani reviewed (commented):
This seems like a helpful addition! I left a few high-level comments to start.

One overall comment: the original issue mentions exposing an iterator, whereas we return a decoded array here. I don't have major concerns with this, but wanted to highlight the difference.

mayya-sharipova committed:
Address Julie's feedback

-decodeVectorMagnitude go back to Buffer
-correct documentation to use doc access methods for
	vector values and magnitude instead of get functions
- rename getVectorMagnitude to getMagnitude
- remove unnecessary yml tests

mayya-sharipova commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/vectors/20_dense_vector_special_cases.yml:
```diff
@@ -189,3 +189,55 @@ setup:
 - match: {hits.hits.1._score: 0.0}
+
+---
+"No sort, no aggs, no docvalue_fields are allowed":
```
Great comment, removed in 388b1f85029214d3b8cd53f5f2ed5aec2023f6f3

mayya-sharipova commented on docs/reference/vectors/vector-functions.asciidoc:
```diff
@@ -195,3 +205,50 @@ You can check if a document has a value for the field `my_vector` by
+
+- `float[] getVectorValue()` – returns a vector's value as an array of floats
+
+- `float getVectorMagnitude()` – returns a vector's magnitude (available for
```
@jtibshirani Thanks for the feedback. I was also thinking the same – to implement a slower version of `getVectorMagnitude`, but this would require decoding the whole vector. If a user is already using `vectorValue` in their script and decoding a vector, using `magnitude` would mean decoding this vector the second time.  So it would be faster for a user to implement `magnitude` function themselves. WDYT?

jtibshirani replied:
To me it's the right trade-off for a simple API. First, it will only be slow for vectors indexed before 7.5, which was before they were even GA. It also seems okay that it's slow, users can easily work around it. Maybe we could just write a short note in the docs about the pre-7.5 behavior so that users are aware.

jtibshirani commented on docs/reference/mapping/types/dense-vector.asciidoc:
```diff
@@ -10,10 +10,6 @@ A `dense_vector` field stores dense vectors of float values.
 The maximum number of dimensions that can be in a vector should
 not exceed 2048. A `dense_vector` field is a single-valued field.
 
-These vectors can be used for <<vector-functions,document scoring>>.
```
I'm curious why we removed this paragraph?

mayya-sharipova replied:
It was not removed, just moved down. I put it back to its original place in 34d961b92339750bdd649bcb5e0b2064609fc578.

jtibshirani replied:
Sorry, I read this part too quickly.

jtibshirani commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/vectors/30_dense_vector_script_access.yml:
```diff
@@ -0,0 +1,142 @@
+          - '{}'
+
+  # check getVectorValue() API
+  - do:
```
This test coverage looks good! Since unit tests are generally easier to work with than REST tests, I wondered if there was a way perform some of the same checks as unit tests?

mayya-sharipova replied:
@jtibshirani Thanks for the feedback. I could not find any good examples of doing unit tests with script doc values.  Unit tests that we have they mock scripts contexts and mock what script returns, which kind of defeats the purpose of testing what `getVectorValue()` and `getMagnitude()` returns.

I am happy to redesign tests as unit tests, if you know any examples I can follow.

jtibshirani replied:
It seems like there are some simple cases where we just want to check `getVectorValue` and `getMagnitude` return the right value or error appropriately. These could be covered in a test like `DenseVectorScriptDocValuesTests`. A similar test would be `ScriptDocValuesGeoPointsTests`.

jtibshirani commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/vectors/30_dense_vector_script_access.yml:
```diff
@@ -0,0 +1,142 @@
+  # check getMagnitude() API
+  - do:
+      headers:
+        Content-Type: application/json
```
I didn't have the chance to dig into it, but I was curious why we use this header here? I noticed it because we have a special skip for `features: headers`.

mayya-sharipova replied:
It looks like some content types can't deal well with floating points numbers in test assertions (e.g. assertions like `gte: {hits.hits.0._score: 3.3166}`), and tests fail with errors: "java.lang.Double cannot be cast to class java.lang.Float".

I am not sure which are these content types (at test failures don't display them), but `Content-Type: application/json` always works well, and its our standard content type.

jtibshirani replied:
Thanks for the explanation.

mayya-sharipova committed:
Address Julie's feedback 2

mayya-sharipova commented:
@jtibshirani Thanks for another round of review. I've tried to address in https://github.com/elastic/elasticsearch/pull/71313/commits/34d961b92339750bdd649bcb5e0b2064609fc578

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into dense-vector-iterator

mayya-sharipova committed:
Address style check

jtibshirani commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/query/ScoreScriptUtils.java:
```diff
@@ -67,18 +70,10 @@ BytesRef getEncodedVector() {
                 throw new IllegalArgumentException("A document doesn't have a value for a vector field!");
             }
-
-            int vectorLength = VectorEncoderDecoder.denseVectorLength(scoreScript._getIndexVersion(), vector);
```
It's great we were able to remove these references to `ScoreScript#_getIndexVersion`. I think this method can now be deleted?

mayya-sharipova replied:
Great point! Addressed in a66f75590ba961439ac02907a46074312d3a80a2

jtibshirani commented on docs/reference/vectors/vector-functions.asciidoc:
```diff
@@ -195,3 +205,51 @@ You can check if a document has a value for the field `my_vector` by
+- `doc[<field>].vectorValue` – returns a vector's value as an array of floats
+
+- `doc[<field>].magnitude` – returns a vector's magnitude as a float
+(for vectors created prior version 7.5 magnitude is not stored.
```
Small corrections: "for vectors created prior **to** version 7.5 **the** magnitude is not stored. So this function calculates it anew every time **it** is called"

mayya-sharipova replied:
Addressed in a66f75590ba961439ac02907a46074312d3a80a2

mayya-sharipova committed:
Address Julie's feedback 3

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into dense-vector-iterator

mayya-sharipova commented:
@jtibshirani  Thanks for another round of review. I've tried to address your feedback in a66f75590ba961439ac02907a46074312d3a80a2

About `DenseVectorScriptDocValuesTests`, that's a great feedback.
I've added this test class to test basic functionalities of `DenseVectorScriptDocValues`.

I've still kept `30_dense_vector_script_access.yml ` for a more involved test case using an array of query vectors to test to original problem reported by a user in https://github.com/elastic/elasticsearch/issues/70437

jtibshirani reviewed (approved):
Looks good to me.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into dense-vector-iterator

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into dense-vector-iterator

mayya-sharipova commented:
@elasticmachine check

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/bwc

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into dense-vector-iterator

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/2

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/2

mayya-sharipova commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into dense-vector-iterator
